### ISSUE TITLE
add useBest() function?

### ISSUE THREAD

**Issue by dankelley** [ArgoCanada/argoFloats#408]
We (I and @j-harbin and @richardsc) discussed this.  A rough code idea is as follows.  I won't explain more now but I'll fiddle with this over the weekend.
```R

useBest(a)
{
    if (!inherits(a, "argoFloats"))
        stop("'a' must be an object created with argoFloats::readProfiles()")
    rval <- a
    for (i in seq_along(a[["argos"]])) {
        message("i=", i)
        ai <- a[[i]]
        varNames <- names(ai[["data"]])
        message("varNames: \"", paste(varNames, collapse="\" \""), "\"")
        varNamesRaw <- varNames[!grepl(varNames, "Adjusted")]
        message("varNamesRaw: \"", paste(varNamesRaw, collapse="\" \""), "\"")
        varNamesAdjusted <- varNames[grepl(varNames, "Adjusted")]
        message("varNamesAdjusted: \"", paste(varNamesAdjusted, collapse="\" \""), "\"")
        if ("dataMode" %in% names(ai@metadata)) { # core
            dm <- ai@metadata$dataMode
            # alter (rename etc)
            if (dm == "A") {
                message("adjusted data")
                for (name in varNamesRaw) {
                    adjustedName <- paste0(name, "Adjusted")
                    if (adjustedName %in% varNamesAdjusted) {
                        ai@data[name] <- ai@data[adjustedName]
                    }
                }
                # copy .e.g. data$salinityAdjusted into data$salinity
            } else if (dm == "R") {
                message("realtime data: FIXME: what to do?")
            } else if (dm == "D") {
                message("delayed-mode data")
            } else {
                stop("dataMode must be \"A\", \"R\", or \"D\", not \"", dm, "\"")
            }
        } else if ("parameterDataMode" %in% names(ai@metadata)) { # BGC
            message("handle BGC")
        }
    }
    rval
}

```

**Comment by dankelley** [comment]
Um, flags need to be looked at too, don't they?  Q: is there one flag for e.g. `salinity` and another flag for `salinityAdjusted`?

**Comment by dankelley** [comment]
I've done a test code (below) that yields some results (attached) that are useful in exploring what can be in files.  This code does **not** do all we want; I am just fiddling with ideas, and filling in blanks.

In the attached file, note

* items 1 and 23 (and others) are realtime by filename and realtime by `dataMode`, **but** have
 columns named as adjusted.  How can that be?  The values are all `NA`, though.  (I will write code to check whether we can see non-NA data in such cases.)
* item 24 is realtime by filename but adjusted by `dataMode`, **and** it has non-`NA` data in the adjusted columns.  This seems to mean that the filename does *not* tell us the mode.
* item 48 is realtime by filename and realtime by `dataMode`, **but** it has non-`NA` data for pressure and temperature (but **not** for salinity).

Also, note that I am only looking at some subcases.  I will explore more on the weekend.


[13_useBest_01.txt](https://github.com/ArgoCanada/argoFloats/files/6052273/13_useBest_01.txt)


<details><summary>test R code</summary>

![13_useBest_01.txt](https://github.com/ArgoCanada/argoFloats/files/6052272/13_useBest_01.txt)
```R
library(argoFloats)
set.seed(408)
if (!exists("a")) {
    i <- getIndex(age=10) # no need to get latest and greatest
    n <- 200
    s <- subset(i, sample(seq_len(i[["length"]]), n))
    a <- readProfiles(getProfiles(s))
}

debug <- 1

useBest <- function(a)
{
    if (!inherits(a, "argoFloats"))
        stop("'a' must be an object created with argoFloats::readProfiles()")
    rval <- a
    for (i in seq_along(a[["argos"]])) {
        ai <- a[[i]]
        fn <- ai[["filename"]]
        typeFromFilename <- switch(substring(gsub(".*/","",fn),1,1), "A"="adjusted", "D"="delayed", "R"="realtime")
        message(sprintf("\n%2d. %s (filename suggests %s data)", i, fn, typeFromFilename))
        varNames <- names(ai[["data"]])
        adjustedNames <- grepl("Adjusted$", varNames)
        varNamesRaw <- varNames[!adjustedNames]
        varNamesAdjusted <- varNames[adjustedNames]
        if (debug > 1) {
            message("      varNames: \"", paste(varNames, collapse="\" \""), "\"")
            message("      varNamesRaw: \"", paste(varNamesRaw, collapse="\" \""), "\"")
            message("      varNamesAdjusted: \"", paste(varNamesAdjusted, collapse="\" \""), "\"")
        }
        if ("dataMode" %in% names(ai@metadata)) { # core
            dm <- ai@metadata$dataMode[1]
            message("    core dataset with dataMode=\"", dm, "\"")
            if (dm == "A") {
                for (name in varNamesRaw) {
                    adjustedName <- paste0(name, "Adjusted")
                    if (adjustedName %in% varNamesAdjusted) {
                        nok <- sum(is.finite(ai@data[[adjustedName]]))
                        ai@data[[name]] <- ai@data[[adjustedName]]
                        message("      copied ", adjustedName, " into ", name, " (had ", nok, " finite adjusted data)")
                    }
                }
                # FIXME: copy flags also
            } else if (dm == "R") {
                for (name in varNamesRaw) {
                    adjustedName <- paste0(name, "Adjusted")
                    if (adjustedName %in% varNamesAdjusted) {
                        nok <- sum(is.finite(ai@data[[adjustedName]]))
                        ai@data[[name]] <- ai@data[[adjustedName]]
                        message("      copied ", adjustedName, " into ", name, " (had ", nok, " finite adjusted data)")
                    }
                }
             } else if (dm == "D") {
                message("    delayed-mode (FIXME: code more)")
            } else {
                stop("dataMode must be \"A\", \"R\", or \"D\", not \"", dm, "\"")
            }
        } else if ("parameterDataMode" %in% names(ai@metadata)) { # BGC
            dm <- ai@metadata$dataMode
            message("   BGC dataset (contains \"parameterDataMode\"=\"", paste(dm, collapse="\" \""), "\")")
        }
    }
    rval
}

b <- useBest(a)
```

**Comment by dankelley** [comment]
Note: the above assumes that a given file has only one `DATA_MODE` value, but the following illustrates that this is not the case. (This does not contradict the results shown above; it just says that the code will have to work through the columns individually.)

<details><summary>Demonstration that dataMode can be a vector</summary>

``` r
library(ncdf4)
library(oce)
#> Loading required package: gsw
#> Loading required package: testthat
#> Loading required package: sf
#> Linking to GEOS 3.8.1, GDAL 3.1.4, PROJ 6.3.1
library(curl)
#> 
#> Attaching package: 'curl'
#> The following object is masked from 'package:oce':
#> 
#>     curl

# Download file locally, independently of argoFloats
url <- "https://data-argo.ifremer.fr/dac/aoml/5904013/profiles/D5904013_153.nc"
file <- gsub(".*/", "", url)
if (!file.exists(file))
    curl_download(url, file)

# Show mode, checking read.oce() vs direct ncdf4 lookup
n <- nc_open(file)
ncvar_get(n, "DATA_MODE")
#> [1] "DAD"
d <- read.oce(file)
d[["dataMode"]]
#> [1] "D" "A" "D"
```

<sup>Created on 2021-02-27 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by dankelley** [comment]
I have been looking into this for some cases.  In particular, https://github.com/ArgoCanada/argoFloats/blob/develop/sandbox/dek/13_useBest/13_useBest_02_a.Rmd looks at cycle 153 of float 4904013, which has 3 columns.  The `DATA_MODE` is `"DAD"`, meaning that the first and last columns are "delayed mode" but that the second one is "adjusted mode".

It is worth building the PDF for that Rmd file (type "make" to do that) and examining the last two diagrams.

My question at this point -- motivated by reviewer 3 of the submitted FMS paper -- is whether we ought to apply **separate** rules for individual columns.  Since there's no way, really, to interact with an anonymous reviewer.(they said "yes" to reviewing, but not "yes" to getting into protracted anonymous discussions with us), it is clear that we need to find something in documentation to tell us what to do.  This follows up on the many meetings we had last week.

@j-harbin -- it would be great if you could look through the official Argo documentation to find out just what is considered "proper".  My guess, at this moment, is that we ought to pick and choose between columns of the raw and adjusted matrices individually, according to the corresponding entry in `dataMode`.

In the meantime, I am now looking at a different case (D6902829_047) which has `DATA_MODE` of `"DRRRRRR"`.  The analysis code I wrote for 4904013_053 is not working for this yet.  I'll work on that today, most likely.

**Comment by j-harbin** [comment]
I'll look into this now, Dan. Thanks for the detailed updated.

**Comment by j-harbin** [comment]
The Argo user's manual (Carvel et al. 2019) tells us that `N_PROF` is the number of profiles in a cycle (remember multiple profiles mean multiple columns).  **Section 2.2.5 Measurements for each profile** on page 21-22 gives us all of the answers we are looking for.

When N_PROF > 1, DATA_MODE for each profile can be assigned differently. This is because when there are multiple profiles, delayed-mode or near real-time adjustments can become available at different times.
The adjusted section for each N_PROF should then be filled independently according to its DATA_MODE.
For example, in a profile file with 2 profiles, it is possible that
● DATA_MODE = ’D’ in N_PROF = 1, and
● DATA_MODE = ’R’ in N_PROF = 2.
In this case:
● the adjusted section in N_PROF=1 with DATA_MODE=’D’ should be filled with their adjusted
values;
● the adjusted section in N_PROF=2 with DATA_MODE=’R’ should be filled with FillValue.

In short, I believe each column (or profile) should be handled separately. 








Carval, Thierry, Bob Keeley, Yasushi Takatsuki, Takashi Yoshida, Stephen Loch Loch, Claudia Schmid, and Roger Goldsmith. Argo User’s Manual V3.3. Ifremer, 2019. https://doi.org/10.13155/29825.

**Comment by dankelley** [comment]
I'm closing this, because we have decided through the course of multiple meetings, that this function is not needed at the moment. 

Perhaps we'll resurrect the idea eventually, if there is a need, but the new `useAdjusted()` and `subset(argos,dataMode="")` methods seem to do what users will need, and our goal is to help users, not to fill out some hyperspace of "possible" feature worlds.
